### ISSUE TITLE
Introduction

### ISSUE THREAD

**Issue by kieranpotts** [craftingpress/crafting-css#3]
# Introduction

Cascading Style Sheets (CSS) is a language for describing the presentation of a document, commonly HTML.

***

As the name suggests, Cascading Style Sheets set styles in a cascading fashion. Multiple style rules may be applicable to any one element in the HTML markup. Which rules get applied is determined by cascading down through every ruleset in every style sheet looking for matches. The more specific rulesets override the more general ones. In addition, some styles will cascade down through the DOM and apply themselves to all of an element's descendants, too.

This inheritance model is intrinsic to CSS and cannot be turned off.

In simple, small-scale user interfaces the style cascade can be hugely beneficial. We can set a property just once and have it trickle through to every relevant part of our design. The result: small file sizes, efficient rendering, and highly consistent user interfaces.

But the style cascade can also be a bit of nuisance. The CSS inheritance model tightly couples the presentation of every component of a web page, and that makes it difficult to change stuff. Every change has the potential to inadvertently alter the presentation of other things. We end up frequently resetting bunches of inherited properties that we don't want, which rather negates the benefits of the cascade.

The bigger the project, the more the style cascade is a hindrance than a help. It's why CSS is notoriously difficult to maintain in high-fidelity designs \(where there is lots of detail and variation in the presentation of things\), in long-running projects with many collaborators, and in projects where the team members have widely differing specialities and abilities or where there is a high staff turnover. In these scenarios if everyone is not on the same page in terms of how to structure and manage the project's CSS assets, a codebase can rapidly snowball into an unmanageable mess.

CSS does not scale at all well for today's richly formatted, rapidly iterated websites.

Extended CSS languages – Sass, Less, Stylus, et al – make things a bit better. And some of their best ideas – like variables – are in the processing of being ported to native CSS. But extended CSS languages don't truly fix the scalability issue. They just skirt around the problem by encouraging code duplication, deep nesting of selectors and other bad ideas.

What we really need is a proper scoping mechanism, a standardised way to encapsulate styles within fragments of markup. This would allow us to split up complex designs into lots of bite-size modules of HTML, CSS. and optionally JavaScript. Each module could be developed and maintained separately of the rest. Team members could more easily work on different things in parallel. And it would be easier to take UI components from previous projects and plug them straight in to new projects, speeding up development cycles.

Modular CSS would be the ultimate scalable CSS.

There is hope. The Shadow DOM – part of the exciting Web Components standard dreamt up at the W3C – will encapsulate styles by default. A new `@scope` rule is at an early proposal stage. And the `scoped` attribute, proposed for the `<style>` element, will limit the application of styles to the part of the DOM tree where the `<style>` element is located.

```
<div>
    <style scoped>
        h2 { 
            color: red; 
        }
    </style>
    <h2>This is red</h>
</div>

<h2>This is NOT red</h2>
```

In addition, soon we'll be able to use a new property called "all" to fully reset things back to their default presentation, effectively restarting the inheritance chain.

```
*.reset-me { 
    all: unset; 
}
```

None of these things are useful right now, because they do not yet work consistently in all mainstream browsers. The Shadow DOM is implemented only in Chrome, Opera and Android 4.4's native browser, and it is an experimental feature in Firefox. The `all` property works only in recent editions of Chrome, Firefox, Opera and Android 4.4.4+. As for scoped CSS, its future is highly uncertain. The `scoped` attribute showed up briefly in Chrome \(behind a flag\) but subsequently vanished, and Microsoft says it has "no current plans" to implement scoped styles in Edge. At this time scoped styles work only in Firefox.

No matter. Things are certainly moving in the direction of modularity. In the future we will have standardised methods of scoping presentational styles to particular bits of an HTML document and locking out the inheritance chain from others. This will make it possible to develop web interfaces as a series of stand-alone UI components. And that will make it much easier to manage the codebases of large-scale web projects.

In the meantime we must find a way to do this ourselves. One option would be to polyfill Web Components in browsers that lack native support \([http://webcomponents.org/polyfills/\)](http://webcomponents.org/polyfills/%29\). Personally, I'm not keen on the idea of building production-ready web apps on emerging technologies that are still in a state of flux. There is still disagreement among browser vendors over implementation of some aspects of Web Components, such as HTML imports which are effectively on hold due to the overlap with ECMAScript 6 module loading. Furthermore, polyfills are heavy and the correct rendering of baseline content becomes dependent upon JavaScript. Bespoke solutions such as CSS Modules \([https://github.com/css-modules](https://github.com/css-modules%29%29\) have the same problem.

Right now, the best option is to devise our own system to manage CSS at scale.

# Web design methodologies

A web design methodology is a system for applying CSS to HTML. The primary aim is to make it possible to break up big, complex web designs into lots of smaller, simpler modules of HTML and CSS \(and optionally JavaScript\) that can be developed and maintained independently of one another. But a good web design methodology will achieve much more. It will:

* Make the design process faster.
* Make front-end code easier to read and understand.
* Make it easier for multiple people to collaborate on a design.
* Make for a smoother transition in the handover of a project to a new team.
* Reduce the learning curve for new designers joining a project.
* Help developers understand the relationship between HTML and CSS in a project.
* Make it easier to change things.
* Make it easier to scale a design to accommodate more content and functionality.
* Make it easier to add more variety to the visual presentation.
* Make it easier to reuse things that already exist in a project codebase.
* Make it easier to port UI components from a previous project into a new one.
* Encourage production of well-formed, accessible, standards-compliant code.
* Enforce consistency and limit duplication and bloat.
* Reduce page size and increase page rendering speed.
* Provide ready-made documentation.
* Increase productivity.
* Make web design a more rewarding, enjoyable experience.

Web design methodologies benefit all kinds of web projects, big and small. They benefit freelancers who work solo on small-scale web projects. And they are absolutely essential for large and long-running projects.

A few entrepreneurial types have developed their web design methodologies into formal design systems and published them for everyone's benefit. Nicole Sullivan's **Object-Oriented CSS Framework** \([https://github.com/stubbornella/oocss/](https://github.com/stubbornella/oocss/%29%29\), published in 2009, was the first web design methodology to become widely adopted. It is still hugely influential. Jeremy Clarke's **Don't Repeat Yourself** principles \([https://vimeo.com/38063798](https://vimeo.com/38063798%29%29\), presented at ConFoo in 2012, has similar goals to OOCSS, but offers a very different solution. Brad Frost followed up with a widely circulated blog post titled **Atomic Design** in 2013 \([http://bradfrostweb.com/blog/post/atomic-web-design](http://bradfrostweb.com/blog/post/atomic-web-design%29%29\), which got us thinking about web interfaces as a hierarchy of progressively more complex objects built up from lots of smaller ones. Other popular web design methodologies include: **Block Element Modifier** \([http://bem.info/](http://bem.info%29%29\), developed by the dev team at Yandex; Jonathan Snook's **Scalable and Modular Architecture for CSS** \([http://smacss.com/](http://smacss.com%29%29\); and Nicolas Gallagher's **SUIT CSS** \([http://suitcss.github.io/](http://suitcss.github.io%29%29\), which combines a class naming system with extended CSS syntax.

# About this book

TODO...

This book presents a brand new web design methodology – mine. This is how I write and manage CSS at scale. I call it Systematic Web Design. It is a solution that I have developed over a period of at least three years. It is influenced by many of the ideas and conventions of existing web design methodologies, though the practical implementation is generally simpler. There is less to remember.

Systematic Web Design – or just "Systematic" for short – is a pure CSS methodology, though it works equally well whether you are writing Sass or Less or any other superset of CSS. Personally I use only a limited amount of extended CSS syntax, just variables and a few mixins. Most of the other features of extended CSS languages become redundant once you've got a solid web design methodology in place.

Systematic Web Design prescribes strict class naming conventions, which reflect how this design system separates out the various concerns in a user interface into discrete units of code. But that's all it does. Systematic Web Design does not wander into the territory of guidelines, build tools or workflow processes. The coding style you adopt in writing CSS, and how you comment and document your code, is left up to you. Style guides, UI pattern libraries, tools like preprocessors and linters, frameworks like Bootstrap and Skeleton, and formal workflow systems like Scrum and pair programming – all of these things can compliment the practical implementation of a web design methodology like Systematic, but these things are outside of the scope of this book.

The first chapter of this book summarises the software design principles – things like encapsulation, loose coupling and defensive programming – that have influenced the Systematic Web Design methodology. The next four chapters are dedicated to the four component parts of Systematic Web Design: elements, layout, widgets and modifiers. There follows a chapter that summarises Systematic's class naming conventions, and a short closing chapter deals with the practicalities of applying Systematic Web Design in the context of responsive design and CSS preprocessors.

There is no "best" web design methodology. Different approaches work better for different types of projects. What matters more than the specifics of any particular web design methodology is the consistency with which one is implemented within a codebase. I can't encourage you to religiously execute my web design methodology in every project when I don't even do that myself. But this is always my starting point. Systematic Web Design is elementary enough that it can be easily tweaked and extended on a project-by-project basis, moulded to better fit the skill set and creative temperament of the wider team.

Web design methodologies have a long future. Even with standard systems for modular UI design on the horizon, there will always be a need for sober, articulate front-end architecture.

***

TODO:

Even among the pros there is little agreement over the "best way" to write CSS. Ask a bunch of web designers to implement a mockup and they will each produce completely different code. That's because there isn't really a right or wrong way of doing CSS. It is a very loose language. How we write CSS becomes a very personal expression of our creative principles. And that is why discussion of CSS coding conventions and best practices stokes such heated argument in the web design community.
